nucypher/blockchain/eth/deployers.py

```diff
@@ -151,26 +154,149 @@ def _ensure_contract_deployment(self) -> bool:
         Emits the configured blockchain network transactions for single contract instance publication.
         """
-        # TODO: Consider splitting this into to classes upgradeable v non-upgraeable.
-        #       Alternately, just remove deployment secrets alltogether.
         raise NotImplementedError
 
     def make_agent(self) -> EthereumContractAgent:
         agent = self.agency(registry=self.registry, contract=self._contract)
         return agent
 
 
-class NucypherTokenDeployer(ContractDeployer):
+
+class OwnableContractMixin:
+
+    _ownable = True
+
+    class NotOwnable(RuntimeError):
+        pass
+
+    def transfer_ownership(self, new_owner: str, transaction_gas_limit: int = None):
+        if not self._ownable:
+            raise self.NotOwnable(f"{self.contract_name} is not ownable.")
+
+        # Get Bare Contracts
+        existing_bare_contract = self.get_latest_version(registry=self.registry,
+                                                         provider_uri=self.blockchain.provider_uri)
+
+        proxy_deployer = self._proxy_deployer(registry=self.registry,
+                                              target_contract=existing_bare_contract,
+                                              deployer_address=self.deployer_address,
+                                              bare=True)  # acquire agency for the dispatcher itself.
+
+        # Stage
+        contract_function = existing_bare_contract.functions.transferOwnership(new_owner)
+        proxy_contract_function = proxy_deployer.contract.functions.transferOwnership(new_owner)
+
+        # Execute
+        principal_receipt = self.blockchain.send_transaction(sender_address=self.deployer_address,
+                                                             contract_function=contract_function,
+                                                             transaction_gas_limit=transaction_gas_limit)
+
+        proxy_receipt = self.blockchain.send_transaction(sender_address=self.deployer_address,
+                                                         contract_function=proxy_contract_function,
+                                                         transaction_gas_limit=transaction_gas_limit)
+
+        # Report
+        receipts = {'principal': principal_receipt, 'proxy': proxy_receipt}
+        return receipts
+
+
+class UpgradeableContractMixin:
+
+    _upgradeable = True
+    _proxy_deployer = NotImplemented
+
+    class NotUpgradeable(RuntimeError):
+        pass
+    
+    def deploy(self, secret_hash: bytes, gas_limit: int, progress) -> dict:
+        """
+        Provides for the setup, deployment, and initialization of ethereum smart contracts.
+        Emits the configured blockchain network transactions for single contract instance publication.
+        """
+        if not cls._upgradeable:
+            raise cls.NotUpgradeable(f"{cls.contract_name} is not upgradeable.")
```

Comment:
How come you used `cls` instead of `self` here?

Reply:
Accessing class variables and not instance variables. :)

https://www.digitalocean.com/community/tutorials/understanding-class-and-instance-variables-in-python-3

Reply:
Haha sure, however, in other spots we use `self` instead of `cls`, example in the `upgrade()` and `rollback()` functions.

Reply:
Oh, then that's probably a legit issue, I think.

Reply:
Good eye! :heavy_check_mark: